### ISSUE TITLE
Potential issue with gold standards (or fifty shades of myelin)

### ISSUE THREAD

**Issue by perone** [axondeepseg/axondeepseg#139]
I was experimenting with a model and I used data from the folder `SEM_3c_512` (Duke) and I found that the gold standards contained a range of pixel intensities that weren't one of the three classes [background, myelin, axon]. For instance, if I just want the myelin content, I threshold for the myelin pixel intensity and it creates a ground truth like this one below:

![image](https://user-images.githubusercontent.com/412328/47664983-8348a480-db76-11e8-9724-a90413e60a19.png)

![image](https://user-images.githubusercontent.com/412328/47665092-ba1eba80-db76-11e8-9ab7-c6241c1fb360.png)

Where we can see the noisy pixels on many places. Please note that GitHub _is resampling these images_, so they look less bad then they're in reality.

When we look on the original gold standards, we can see that there are indeed many different shades representing the classes:

![image](https://user-images.githubusercontent.com/412328/47665213-fce09280-db76-11e8-88d0-c370ea72fa52.png)

This might be an effect of resampling of the gold standards, that might be creating interpolation on the gold standard and therefore messing up with the class definitions. This may be a problem for AxonDeepSeg if the training is being done without taking into consideration these issues, but I'm not sure how these gold standars are being handled today on AxonDeepSeg. I'm not sure if the data used for training is also from this same folder on Duke.

**Comment by mathieuboudreau** [comment]
I suspect this line could be the source of this issue:

https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/dataset_building.py#L48

I'm not 100% sure, but by quickly browsing Duke, it appears to me that this issue only happens to patched datasets, not the raw masks. I can't confirm though, as I am not sure yet where the matching raw datasets from the "gold standard" are located. @perone would a nearest-neighbour interpolation be the best approach? Currently we're using the default (`order=1` -> bi-linear), whereas setting `order=0` would use a nearest neighbour interpolation.

Bonus: these lines could also be creating to a similar problem that's hidden from the end-user during training/data-augmentation:

https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/data_augmentation.py#L79

https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/data_augmentation.py#L139

https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/data_augmentation.py#L181-L182

I'm in the process of writing a CLI interface that might be useful for pre-training quality control of the datasets in branch [`quality_control`](https://github.com/neuropoly/axondeepseg/tree/quality_control).

**Comment by mathieuboudreau** [comment]
Just did a quick test of raw_img_to_patches. Downsampling (e.g. 0.07 to 0.1) doesn't seem to cause introduce new values (or maybe very few), whereas upsampling (e.g. 0.37 to 0.1) predictably introduces almost every possible bin value into the mask. Setting `order=0` seems to resolve this.

 So this is definitely a big part of this problem, but it may be present elsewhere (e.g. data augmentation).

**Comment by mathieuboudreau** [comment]
Here's a gif of a patch generated with `order=1` (bi-linear interpolation) vs `order=0` (nearest neighbour interpolation).

![order_1_v_0](https://user-images.githubusercontent.com/1421029/47886145-61bc1700-de1b-11e8-9d1d-72b5cd00b745.gif)

**Comment by vs74** [comment]
> I suspect this line could be the source of this issue:
> 
> https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/dataset_building.py#L48
> 
> I'm not 100% sure, but by quickly browsing Duke, it appears to me that this issue only happens to patched datasets, not the raw masks. I can't confirm though, as I am not sure yet where the matching raw datasets from the "gold standard" are located. @perone would a nearest-neighbour interpolation be the best approach? Currently we're using the default (`order=1` -> bi-linear), whereas setting `order=0` would use a nearest neighbour interpolation.
> 
> Bonus: these lines could also be creating to a similar problem that's hidden from the end-user during training/data-augmentation:
> 
> https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/data_augmentation.py#L79
> 
> https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/data_augmentation.py#L139
> 
> https://github.com/neuropoly/axondeepseg/blob/9e59f536da6c709d0341999ae2b11945de588cb2/AxonDeepSeg/data_management/data_augmentation.py#L181-L182
> 
> I'm in the process of writing a CLI that might be useful for pre-training quality control of the datasets in branch [`quality_control`](https://github.com/neuropoly/axondeepseg/tree/quality_control).
> 
> Edit: I should also write tests in the testing suite to ensure that masks after splitting into patches or dataset augmentation preserve the number and value of unique pixel values. The new function that I wrote [here](https://github.com/neuropoly/axondeepseg/blob/f99d52693b3e15359aec1b36580d17d217fb75bc/AxonDeepSeg/visualization/get_masks.py#L68-L90) should help.

@ Mathieu, could you check if the problem still persists by running script on raw data to convert them into patches ? As the problem lies with gold standard so that could be the probable cause. Regarding data augmentation is concerned it's accurately handling this issue during training (current ADS) so no issue on this side.

**Comment by vs74** [comment]
One more point I would like to bring into notice for the `gold standard` that is there should be no background in between `myelin` and` axon`. I believe it should be either `axon` or `myelin`. Some of the gold standard here are having this problem. 
![Screen Shot 2019-07-30 at 5 31 20 PM](https://user-images.githubusercontent.com/31419210/62167003-2db14c00-b2f0-11e9-9047-2575dc60b503.png)
What do you think @mathieuboudreau / @jcohenadad  is it a problem or is it fine ?

**Comment by jcohenadad** [comment]
> What do you think @mathieuboudreau / @jcohenadad is it a problem or is it fine ?

it is a problem

**Comment by vs74** [comment]
> > What do you think @mathieuboudreau / @jcohenadad is it a problem or is it fine ?
> 
> it is a problem
So, I have 2 questions in this regard : -
1. Should we handle it in Dataset building(from raw datasets to patches) or during training ?
2. Should we replace this value either by **myelin** or by **axon** values else should go we with different approach ?

**Comment by jcohenadad** [comment]
1. We should fix the gold standard at the first place
2. The answer is in the raw data (ie: the image)

**Comment by mathieuboudreau** [comment]
Stale - was resolved before (eg in #242 and others)
